Sources/_MatchingEngine/Regex/Parse/LexicalAnalysis.swift

```diff
@@ -633,14 +661,140 @@ extension Source {
+  ///
+  ///     EscapedReference -> 'g{' NameOrNumberRef '}'
+  ///                       | 'g<' NameOrNumberRef '>'
+  ///                       | "g'" NameOrNumberRef "'"
+  ///                       | 'g' NumberRef
+  ///                       | 'k<' <String> '>'
+  ///                       | "k'" <String> "'"
+  ///                       | 'k{' <String> '}'
+  ///                       | [1-9] [0-9]+
+  ///
+  private mutating func lexEscapedReference(
+    numberOfPriorGroups: Int
+  ) throws -> Located<AST.Atom.Kind>? {
+    try recordLoc { src in
+      if src.tryEat("g") {
+        // PCRE-style backreferences.
+        if src.tryEat("{") {
+          let ref = try src.expectNamedOrNumberedReference(
+            endingWith: "}").value
+          return .backreference(ref)
+        }
+
+        // Oniguruma-style subpatterns.
+        if let openChar = src.tryEat(anyOf: "<", "'") {
+          let ref = try src.expectNamedOrNumberedReference(
+            endingWith: String(Source.getClosingDelimiter(for: openChar))).value
+          return .subpattern(ref)
+        }
+
+        // PCRE allows \g followed by a bare numeric reference.
+        if let ref = try src.lexNumberedReference() {
+          return .backreference(ref.value)
+        }
+
+        // Fallback to a literal character. We need to return here as we've
+        // already eaten the 'g'.
+        return .char("g")
+      }
+
+      if src.tryEat("k") {
+        // Perl/.NET-style backreferences.
+        if let openChar = src.tryEat(anyOf: "<", "'", "{") {
+          let closingChar = Source.getClosingDelimiter(for: openChar)
+          return .backreference(.named(
+            try src.lexUntil(eating: closingChar).value))
+        }
+        // Fallback to a literal character. We need to return here as we've
+        // already eaten the 'k'.
+        return .char("k")
+      }
+
+      // Lexing \n is tricky. In PCRE it's treated as a backreference if its
+      // first digit is not 0 (as that is always octal) and one of the following
+      // holds:
+      //
+      // - It's 0 < n < 10 (as octal would be pointless here)
+      // - Its first digit is 8 or 9 (as not valid octal)
+      // - There have been as many prior groups as the reference.
+      //
+      // Oniguruma follows the same rules except the second one. e.g \81 and \91
+      // are instead treated as literal 81 and 91 respectively.
+      // TODO: If we want a strict Oniguruma mode, we'll need to add a check
+      // here.
+      if src.peek() != "0", let digits = src.peekPrefix(\.isNumber) {
+        // First lex out the decimal digits and see if we can treat this as a
+        // backreference.
+        let num = try Source.validateNumber(digits.string, Int.self, .decimal)
+        if num < 10 || digits.first == "8" || digits.first == "9" ||
+            num <= numberOfPriorGroups {
+          src.advance(digits.count)
+          return .backreference(.absolute(num))
+        }
```

Comment:
Should we just return the octal literal?

Reply:
We can do, though we'd need to call into `expectUnicodeScalar` as we can only read up to 3 digits of octal, and it wouldn't cover the `\0nn` case unless we also add a condition for that here

Reply:
Where is that logic when we fall through?

Reply:
In `expectUnicodeScalar`, called from `expectEscaped`

Reply:
It seems more straight-forward to make the explicit call, especially if we have context-local constraints on the number of digits or other aspects of interpretation. Otherwise, would the general case have to reconstruct this logic?

Fall-through also means that any intermediary code may have to be concerned with or reason about this possibility. In general I'm a fan of local reasoning. But this can be done after this PR, and it's not a huge deal.

Reply:
> Otherwise, would the general case have to reconstruct this logic?

I might not be following what you mean by this, but I don't *think* so. Both the backreference and unicode scalar cases have differences in what they accept, but there is a syntactic ambiguity between them. In general, we just need to make sure to try lex as a backreference before a unicode scalar. IMO if we change this logic to produce a scalar, we should also change the scalar logic to be able to produce a backreference so we no longer need to reason about which is called first. I'm happy to do that in a follow-up, but going to merge this for now to cleanup the option parsing PR